packages/lodestar/src/chain/blocks/pool.ts

```diff
@@ -1,81 +1,128 @@
 import {SignedBeaconBlock, Slot} from "@chainsafe/lodestar-types";
-import {toHexString} from "@chainsafe/ssz";
 
-import {IBlockProcessJob} from "../interface";
-import {ChainEventEmitter} from "../emitter";
-import {IForkChoice} from "@chainsafe/lodestar-fork-choice";
-import {findUnknownAncestor} from "./util";
+import {IBlockJob} from "../interface";
+import {BlockProcessor} from "./processor";
 
+/**
+ * The BlockPool is a cache of blocks that are pending processing.
+ *
+ * Pending blocks come in two varieties:
+ * - blocks with unknown parents
+ * - blocks with future slots
+ */
 export class BlockPool {
-  private unknownParentBlockPool = new Map<string, IBlockProcessJob[]>();
-  private pendingSlotBlockPool = new Map<Slot, IBlockProcessJob[]>();
+  /**
+   * Blocks indexed by blockRoot
+   */
+  public blocks: Map<string, IBlockJob>;
+  /**
+   * Blocks indexed by parentRoot, then blockRoot
+   */
+  public blocksByParent: Map<string, Map<string, IBlockJob>>;
+  /**
+   * Blocks indexed by slot, then blockRoot
+   */
+  public blocksBySlot: Map<Slot, Map<string, IBlockJob>>;
 
   private readonly config: IBeaconConfig;
-  private readonly blockProcessorSource: Pushable<IBlockProcessJob>;
-  private readonly eventBus: ChainEventEmitter;
-  private readonly forkChoice: IForkChoice;
+  private readonly processor: BlockProcessor;
 
-  constructor(
-    config: IBeaconConfig,
-    blockProcessorSource: Pushable<IBlockProcessJob>,
-    eventBus: ChainEventEmitter,
-    forkChoice: IForkChoice
-  ) {
+  constructor({config, processor}: {config: IBeaconConfig; processor: BlockProcessor}) {
     this.config = config;
-    this.blockProcessorSource = blockProcessorSource;
-    this.eventBus = eventBus;
-    this.forkChoice = forkChoice;
+    this.processor = processor;
+
+    this.blocks = new Map();
+    this.blocksByParent = new Map();
+    this.blocksBySlot = new Map();
   }
 
-  public addPendingBlock(job: IBlockProcessJob): void {
-    const key = this.getKey(job.signedBlock);
-    const pendingBlockPool = this.unknownParentBlockPool.get(key);
-    if (pendingBlockPool) {
-      pendingBlockPool.push(job);
-    } else {
-      this.unknownParentBlockPool.set(key, [job]);
-      //this prevents backward syncing, tolerance is 20 blocks
-      if (job.signedBlock.message.slot <= this.forkChoice.getHead().slot + 20) {
-        const pendingBlocks = Array.from(this.unknownParentBlockPool.values()).flat();
-        this.eventBus.emit(
-          "unknownBlockRoot",
-          findUnknownAncestor(this.config, pendingBlocks, job.signedBlock.message.parentRoot)
-        );
-      }
+  public addByParent(job: IBlockJob): void {
+    const {signedBlock} = job;
+    // put block in two indices:
+    // blocks
+    const blockKey = this.getBlockKey(signedBlock);
+    this.blocks.set(blockKey, job);
+    // blocks by parent
+    const parentKey = this.getParentKey(signedBlock);
+    let blocksWithParent = this.blocksByParent.get(parentKey);
```

Comment:
@wemeetagain seems like you missed this one

Reply:
I think this should still be called `blocksWithParent`, or perhaps `blocksAtParent`.
These blocks all _do_ have a parent, its the key that they're all indexed by.

Reply:
Yeah but they are indexed by it because it's missing. I find current name quite confusing though

Reply:
Think of this module as a _just_ a cache with three indices. blocks by root, parentRoot, slot.
when we get blocks by slot, these aren't blocks _without_ a slot. They're blocks _at_ a slot.
Similarly, when we get a block by root, its a block _with/of_ a root.
Same thinking goes for blocks by parent root.

Reply:
But they are blocks without slot, their slot has not yet come or it has already passed. I don't feel like arguing about that anymore but if I find it confusing and I do know what this is about it's highly likely other contributors would as well. At least put comment explaining it since it's not really self descriptive variable name